modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Main.scala

```diff
@@ -0,0 +1,96 @@
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.kinesis
+
+import cats.effect.{ExitCode, IO, IOApp, Resource, Sync, SyncIO}
+
+import cats.Parallel
+import cats.implicits._
+
+import java.util.concurrent.{Executors, TimeUnit}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.ExecutionContext
+
+import fs2.aws.kinesis.CommittableRecord
+
+import com.snowplowanalytics.snowplow.enrich.common.fs2.Run
+import com.snowplowanalytics.snowplow.enrich.common.fs2.Telemetry
+
+import com.snowplowanalytics.snowplow.enrich.kinesis.generated.BuildInfo
+
+object Main extends IOApp.WithContext {
+
+  // Kinesis records must not exceed 1MB
+  private val MaxRecordSize = 1000000
+
+  /**
+   * An execution context matching the cats effect IOApp default. We create it explicitly so we can
+   * also use it for our Blaze client.
+   */
+  override protected val executionContextResource: Resource[SyncIO, ExecutionContext] = {
+    val poolSize = math.max(2, Runtime.getRuntime().availableProcessors())
+    Resource
+      .make(SyncIO(Executors.newFixedThreadPool(poolSize)))(pool =>
+        SyncIO {
+          pool.shutdown()
+          pool.awaitTermination(10, TimeUnit.SECONDS)
+          ()
+        }
+      )
+      .map(ExecutionContext.fromExecutorService)
+  }
+
+  def run(args: List[String]): IO[ExitCode] =
+    Run.run[IO, CommittableRecord](
+      args,
+      BuildInfo.name,
+      BuildInfo.version,
+      BuildInfo.description,
+      executionContext,
+      DynamoDbConfig.updateCliConfig[IO],
+      Source.init[IO],
+      (_, out, monitoring) => Sink.initAttributed(out, monitoring),
+      (_, out, monitoring) => Sink.initAttributed(out, monitoring),
+      (_, out, monitoring) => Sink.init(out, monitoring),
+      checkpoint[IO],
+      List(_ => S3Client.mk[IO]),
+      getPayload,
+      MaxRecordSize,
+      Some(Telemetry.Cloud.Aws),
+      getRuntimeRegion
+    )
+
+  private def getPayload(record: CommittableRecord): Array[Byte] = {
+    val data = record.record.data
+    val buffer = ArrayBuffer[Byte]()
+    while (data.hasRemaining())
+      buffer.append(data.get)
+    buffer.toArray
+  }
+
+  /** For each shard, the record with the biggest sequence number is found, and checkpointed. */
+  private def checkpoint[F[_]: Parallel: Sync](records: List[CommittableRecord]): F[Unit] =
```

Comment:
This is interesting, because the enrich batch size setting directly affects how often we checkpoint.  If you decrease batch size then you checkpoint more often; therefore might need to bump dynamodb write capacity to compensate.

Probably not a problem for 3.0.0, but definitely something for pipeline operators to be aware of.  Especially as that's not how stream-enrich worked.

Reply:
> batch size setting directly affects how often we checkpoint

I'm not sure to follow, which batch size are you talking about, `maxRecords`?

> might need to bump dynamodb write capacity to compensate

Well I think that if enrich goes beyond DynamoDB write capacity, it really means that enrich is misconfigured.

Reply:
I meant `input.retrievalMode.maxRecords`.

Let's imagine enrich is processing one shard and can process 1000 events per second.

If `input.retrievalMode.maxRecords` is 10000, then enrich will checkpoint once every 10 seconds.  If `input.retrievalMode.maxRecords` is 100, then enrich will checkpoint ten times every seconds.  That might surprise someone that retrievalMode settings can influence checkpoint frequency, unless that person understands the internals of our code.

It's fine for you to resolve this conversation, and not make any changes.  I just think we should remember to document somewhere that `input.retrievalMode.maxRecords` has this slightly surprising consequence on something unrelated to retrieving records.

Reply:
> has this slightly surprising consequence on something unrelated to retrieving records
I'm adding a note in the config

Reply:
https://github.com/snowplow/enrich/pull/481/commits/57f72c1aae548e39f2dd540727f9ca3d5821cfa9